magma/common.py

```diff
@@ -17,6 +18,40 @@ def peek(self):
         return self._stack[-1]
 
 
+class _Ref(object):
+    def __init__(self, value):
+        self.value = value
+
+    def __eq__(self, other):
+        return self.value is other.value
+
+    def __hash__(self):
+        return id(self.value)
+
+
+class IdentitySet(collections.MutableSet):
+    def __init__(self, items=[]):
+        self.refs = set(map(_Ref, items))
+
+    def __contains__(self, elem):
+        return _Ref(elem) in self.refs
+
+    def __iter__(self):
+        return (ref.value for ref in self.refs)
+
+    def __len__(self):
+        return len(self.refs)
+
+    def add(self, elem):
+        self.refs.add(_Ref(elem))
+
+    def discard(self, elem):
+        self.refs.discard(_Ref(elem))
+
+    def __repr__(self):
+        return "%s(%s)" % (type(self).__name__, list(self))
+
+
```

Comment:
Are Python sets deterministically ordered? Otherwise this might mess up the the `__repr__` hashing logic if the ordered of the generated intermediates are different.

Reply:
(e.g. like dictionaries being insertion ordered).

Reply:
Sets are still unordered.
```Python
s0 = set()
# Do a bunch of insertions / deletions from s0
s1 = set(s0) 

assert s0 == s1
# iteration order of a specific set that has not been mutated should be consistent 
# (pretty sure this is not just CPython implementation detail) 
assert list(s0) == list(s0) 
# However this may or may not hold
assert list(s0) == list(s1)
```

Reply:
I think we'll need an ordered set then so we can preserve the uniqueness __repr__ invariant.

Reply:
I should point out that even `dict(...).keys()` / `OrderedDict(...).keys()` do not guarantee iteration for keys generated in different orders, they just guarantee that the same sequence of events will always generate the same order. 

So if you want `__repr__` to be invariant between runs of the same program an insertion ordered structure will do that.  If you want `__repr__` to be invariant between equivalent sets you will need to sort them or something.

Reply:
I know https://pypi.org/project/sortedcontainers/ is popular.

Reply:
Hmm, I think the issue is that two circuit definitions generating the same circuit should be identical, the question is whether generating the set of temporaries will produce the same string (this will be separate code paths generating a set, then comparing the generated strings), so whether those code paths construct the set in the same order I think should be true? So if the same sequence of events generates the same set iteration order, that should be fine.  I don't know if we want to generally sort the temporaries (e.g. circuits defined with the same structure but in different orderings could be viewed as unique or not).  At the end of the day, this is conservative so should not be a problem, but I think if two identical circuits (e.g. an uncached generator) are found, their temporary value sets should iterate the same way so it produces the same repr string for uniqueness checking

Reply:
Good point. Updated to use an ordered version of identity set.

Reply:
An inefficient by quick hack to get an insertion ordered set would be to just wrap a dict.  E.g.:
```
class OSet(MutableSet):
    def __init__(self, init=()):
        self._s = {}
        for elem in init:
            self.add(elem)

    def add(self, elem):
        self._s[elem] = None

    def remove(self, elem):
        del self._s[elem]

    ...
```

Reply:
1. It is necessary for iteration order to be deterministic across runs of the same program/instances of the same circuit. It is good for the keys to be further sorted, but as Lenny pointed out that is not necessary and w/o that guarantee we just have a conservative check. Therefore, all we need is a data-structure whose iteration order is insertion order.

2. @cdonovick: I basically wrapped an ordered dict in `OrderedIdentitySet`. Let me know if the way I did it is a good solution.

3. Will upgrade to `items=()` over `items=[]`.